File: examples/R/sig.R
#Takes a path in R^d given by an array of n points with dims c(d,n)
#and returns its signature up to level m excluding level 0
#e.g. path=array(c(1,1,3,10,20,30),c(3,2))
#Note that the order of path's dimensions is the other 
#way around from iisignature.sig.
sig=function(path,m, flat=TRUE){
  n=dim(path)[2]
  d=dim(path)[1]
  diffs=path[,-1,drop=FALSE]-path[,-n,drop=FALSE]
  if (n<2){
   o=sapply(1:m,function(x)rep(0,d**x))
   if (flat) return (do.call(c,o))
   return (o)
  }
  r=lapply(1:(n-1),function(x)Reduce(kronecker,rep(list(diffs[,x]),m),accumulate=TRUE))
  facts=lapply(1:m,factorial)
  r=lapply(r,function(x)mapply("/",x,facts))
  chen=function(x,y) c(list(x[[1]]+y[[1]]),
                       lapply(2:m,
                              function(z)x[[z]]+y[[z]]
                               +Reduce("+",mapply(kronecker,x[1:z-1],rev(y[1:z-1]),
                                                  SIMPLIFY=FALSE))))
  o=Reduce(chen,r)
  if (flat) return (do.call(c,o))
  o
}

File: examples/matlab/sig.m
%Simple matlab function to match iisignature.sig .
%Returns the signature up to level m of path in R^d 
%given by n points. The path is given as an array
%of shape [n,d].
%Doesn't return level 0, which is always the number 1.
function [sig] = sig(path,m)
    [n,d]=size(path);
    o = {zeros(1,d)};
    for i=2:m
        o{i}=zeros(1,d^i);
    end
    diffs=diff(path,1,1);   
    for i=1:(n-1)
        d=diffs(i,:);
        stepsig={d};
        last=d;
        for j=2:m
            last=kron(d,last)/j;
            stepsig{j}=last;
        end
        for lev=m:-1:1
            o{lev}=o{lev}+stepsig{lev};
            for llev=1:(lev-1)
                o{lev}=o{lev}+kron(o{lev-llev},stepsig{llev});
            end
        end
    end
    %sig=o; %To return a cell array, with the levels separately
    sig=cat(2,o{:}); %To return a single array
end
